### ISSUE TITLE
(Update) Python 2.7 deprecation warning for RHEL 6

### ISSUE THREAD

**Issue by miken32** [certbot/certbot#3726]
Running on RHEL6 (which cannot easily use Python 2.7) this warning message is displayed each time I run the program:

```
/home/foo/.local/share/letsencrypt/lib/python2.6/site-packages/cryptography/__init__.py:26: DeprecationWarning: Python 2.6 is no longer supported by the Python core team, please upgrade your Python. A future version of cryptography will drop support for Python 2.6
  DeprecationWarning
```

This could be easily fixed by changing the shebang in `~/.local/share/letsencrypt/bin/letsencrypt` to the following:

    #!/home/foo/.local/share/letsencrypt/bin/python2 -W ignore::DeprecationWarning

Edit @adferrand: The deprecation warning came back, for python 2.7 from pip execution, since the automated updates of pip in virtualenvs used by certbot-auto.

**Comment by Samureus** [comment]
http://www.curiousefficiency.org/posts/2015/04/stop-supporting-python26.html

**Comment by miken32** [comment]
>if you're currently supporting Python 2.6 for free because folks using RHEL 6 or one of its derivatives say they need it, *and this is proving to be a hassle for you*, then stop.

That's a perfectly reasonable post. However, I don't think that I'm asking for anything too onerous here. If this library stopped working in 2.6 and I was demanding it be fixed, that would be a whole different story. All I want is for a deprecation warning to be silenced, with a single change to a single file.

**Comment by miken32** [comment]
BTW I'd be happy to submit a pull request for this, but I have no idea where to look. The file in question seems to be generated from somewhere else at some point? This codebase is crazy complicated for what seems like a very straightforward task!

**Comment by BenKennish** [comment]
I'm getting this issue too on CentOS 6.8.  It's not much of a problem except for my `certbot-auto renew --quiet` cron job as it means I'm getting an email every time it runs with the warning message.

Is it safe for me to modify the shebang in `~/.local/share/letsencrypt/bin/letsencrypt` as described above or will this change be reverted after an update?

**Comment by BenKennish** [comment]
I modified the shebang as described and, although for a short while I think it was working, I am getting the warning messages in my cron job emails again.  😞

**Comment by miken32** [comment]
I haven't seen any noise from my cron  jobs since making the change. I've searched the source code in vain for where the change needs to be made to submit a PR. No luck yet.

**Comment by BenKennish** [comment]
Oh I've realised that as well as `/home/ben/.local/share/letsencrypt/bin/letsencrypt`, there also exists a `/root/.local/share/letsencrypt/bin/letsencrypt` script on my server.  I guess I need to make the change here too.  

It seems rather untidy that I have two copies.  Is this because one path gets used when I run it as ben (and the script automatically uses `sudo` when necessary) and the other one gets used when I run directly as root (or using sudo from command line)? Can I remove one of these `.local` directories?

**Comment by BenKennish** [comment]
I made the changes to the second script and now it's working.  Thanks @miken32 

I removed all of `/home/ben/.local/share/letsencrypt` and moved my `certbot-auto` script to `/root/bin/` so that I can only ever run it as root and will get consistent results

**Comment by pde** [comment]
Hi!

If you modify `certbot-auto` on your machine you'll need to run it with `--no-self-upgrade` to prevent it from auto-upgrading when a new release is available, and thereby overriding your modification.

The correct way to send us a PR for this would be to modify `letsencrypt-auto-source/letsencrypt-auto.template`.  Look at where we set up to jump from `sh` to Python:

https://github.com/certbot/certbot/blob/master/letsencrypt-auto-source/letsencrypt-auto.template#L331L344

You would want to add that ` -W ignore::DeprecationWarning` Python flag if and only if `$QUIET == 1`. But note that we [aren't naming the python interpreter explicitly](https://github.com/certbot/certbot/blob/master/letsencrypt-auto-source/letsencrypt-auto.template#L340) so you'll have to start doing that, without breaking anything on any platform that this script runs on :)

**Comment by pde** [comment]
Even better: send us a PR that uses one of the RHEL 6 supported mechanisms to install and enable python 2.7.  Then we might just deprecate Python2.6 support across this entire codebase, which would make @alex and the other `python-cryptography` developers very happy.

**Comment by pde** [comment]
(That would be implemented in `letsencrypt-auto-source/pieces/bootstrappers/rpm_common.sh`.

**Comment by miken32** [comment]
@pde You're suggesting changing the call to `$SUDO "$VENV_BIN/letsencrypt" "$@"` but isn't changing the shebang _inside_ `$VENV_BIN/letsencrypt` what we need to do? That's how I've worked around the issue on my machines. Where would I look to modify that file?

As for Python 2.7 on RHEL6, I did it once before but it was from some shady third party repo. I'll poke around and see if I can find anything more reputable!

**Comment by BenKennish** [comment]
If anyone could fix this it'd be great.  I'm getting a bit fed up with manually updating the shebang in `~/.local/share/letsencrypt/bin/letsencrypt` every time certbot-auto updates.  I really don't want to use `--no-self-upgrade`.  Cheers

**Comment by starrychloe** [comment]
Hi this is still occurring

```
/root/.local/share/letsencrypt/lib/python2.6/site-packages/cryptography/__init__.py:26: DeprecationWarning: Python 2.6 is no longer supported by the Python core team, please upgrade your Python. A future version of cryptography will drop support for Python 2.6
  DeprecationWarning

```

**Comment by jelmerjellema** [comment]
I use this in cron:
`/bin/certbot renew -q 2>&1 | /bin/grep -v DeprecationWarning`

**Comment by bmw** [comment]
This has been resolved because `certbot-auto` now requires Python 2.7+ and will walk you through installing Python 3.4 from EPEL on RHEL6 based systems.

**Comment by pixelgaurav** [comment]
now we are getting error for 2.7 going to deprecation please update this

**Comment by adferrand** [comment]
Side effect of the upgrade to pip 19.0 in certbot-auto virtualenv. This could also be fixed again through #6685

**Comment by bmw** [comment]
Cryptography has not deprecated Python 2.7 support which was the cause of the problem in this issue. If you are seeing deprecation warnings, the problem is different and I would encourage you to open a new issue fully describing the problem.
